Add docker dev set up instruction (#3444)
opensearch-project/OpenSearch-Dashboards: docs/docker-dev/docker-dev-setup-manual.md

# Docker Development Environment Setup

> 

Reviewer: I think we need an introductory paragraph that explains why a user might choose this setup, and what following these steps will actually do, at a high level.

Author: Agree, will add one.

> 1. Install [Docker](https://docs.docker.com/get-docker/) if not already.

Reviewer: ```suggestion
1. Install [Docker](https://docs.docker.com/get-docker/) if not already installed.
```

not required, but it can be useful in such cases to provide a quick command to check.

> 2. On the terminal, run command `curl -o- https://raw.githubusercontent.com/opensearch-project/OpenSearch-Dashboards/main/dev-tools/install-docker-dev.sh | bash`. This should create a folder named `opensearch-dashboards-docker-dev` and it should contain two files: `docker-compose.yml` and `entrypoint.sh`.

Reviewer: Other steps use "in the terminal". We should be consistent, and "in" is likely more commonly used

```suggestion
2. On the terminal, run the command `curl -o- https://raw.githubusercontent.com/opensearch-project/OpenSearch-Dashboards/main/dev-tools/install-docker-dev.sh | bash`. This should create a folder named `opensearch-dashboards-docker-dev` and it should contain two files: `docker-compose.yml` and `entrypoint.sh`.
```

We should also provide a plain link to the installer script for people to quick inspect.

Author: Changed the syntax, also added the plain link to the installer script.

> 3. Open [VsCode](https://code.visualstudio.com/download). Install here if not already. Make sure VsCode has the extension `Dev Containers` and `Docker` installed. If not, go to Extensions tab, search and install them.

Reviewer: We may want a little disclaimer at the top of the instructions that state the assumptions (using Docker and VS Code). Otherwise it's a bit surprising to only find out now.

```suggestion
3. Open VS Code or [install it](https://code.visualstudio.com/download), if it's not already. Make sure VS Code has the extensions `Dev Containers` and `Docker` installed. If not, go to Extensions tab, search and install them.
```

Author: Good idea, will add a paragraph at the top indicating this instruction uses docker and vscode.

> 2. In the terminal, run the command `curl -o- https://raw.githubusercontent.com/opensearch-project/OpenSearch-Dashboards/main/dev-tools/install-docker-dev.sh | bash`. This should create a folder named `opensearch-dashboards-docker-dev` and it should contain two files: `docker-compose.yml` and `entrypoint.sh`. Here is the link to the installer script: `https://raw.githubusercontent.com/opensearch-project/OpenSearch-Dashboards/main/dev-tools/install-docker-dev.sh`.

Author: good idea, thx~

4. Under the Discover tab, click `Open Folder`, and open the `opensearch-dashboards-docker-dev` folder that we just created.

> 5. In the workspace folder, set environment variable for the fork repository URL: run `export REPO_URL=[insert your fork repo url here]`. If fork repo has not been created: Go to [OpenSearch Dashboards github page](https://github.com/opensearch-project/OpenSearch-Dashboards) and under fork, select create a new fork, and then copy the https link of the fork url and use it in the above command.

Reviewer: Where/how should they get the fork repo URL? Can they find it from within VS Code?

Reviewer: Also, is this necessary every time they start the container, or is it just for initial setup? Any ways we could automate it?

Author: yea this step for setting up environment variable is necessary every time when they re-start the compose file in a new terminal. Will add a sentence specifying this.

I have explained how to get the fork repo URL in the second part of step 5.

> 6. To run the `docker-compose.yml` file in the background, and type `docker compose up -d --build` in the terminal.

Reviewer: ```suggestion
6. Run the `docker-compose.yml` file in the background by typing `docker compose up -d --build` in the terminal.
```

> 7. Under `Docker` tab in VsCode, verify that there are two containers running: `opensearchproject/opensearch:latest` and `docker.io/library/osd-development`. We can also verify using the command line: `docker ps`.

Reviewer: ```suggestion
7. Under the `Docker` tab in VS Code, verify that there are two containers running: `opensearchproject/opensearch:latest` and `docker.io/library/osd-development`. We can also verify using the command line: `docker ps`.
```

8. Right click `docker.io/library/osd-development`, and select `Attach Visual Studio Code`. This should ssh into the container in another VsCode window.

> 9. For the new VsCode window, if it is not showing the repository code, then select `Open Folder`. Then open `/workspace-docker/OpenSearch-Dashboards`.

Reviewer: ```suggestion
9. For the new VS Code window, if it is not showing the repository code, select `Open Folder`. Then open `/workspace-docker/OpenSearch-Dashboards`.
```

> 10. In the terminal, type  `yarn start:docker` to start the OpenSearch Dashboards application.  (To open a terminal, right click on a file and click `Open in Integrated Terminal`. Then change the terminal type to bash by clicking the + sign on the right top of the terminal window.)

Reviewer: Why do they need to switch to bash?

Author: Yea, thanks for pointing it out, i do not think they have to use a bash terminal.

> 6. Run the `docker-compose.yml` file in the background, and type `docker compose up -d --build` in the terminal.

Reviewer 2: `docker-compose`
